File: addon/Pickup.lua
local _, WS = ...

-- Ground pickups: healing potions, coins, treasure chests, goblin sappers,
-- and lodestones. Everything is pooled and drifts to the survivor once inside
-- pickup range, exactly like XP gems.
WS.Pickup = {}
local Pickup = WS.Pickup

-- `raise` lifts a pickup's draw layer so the important ones never hide behind
-- the sea of coins and XP gems (which sit at the base entity level).
local KINDS = {
    potion = { icon = "Interface\\Icons\\INV_Potion_54", tint = { 1.0, 0.4, 0.4 }, size = 20, raise = 8 },
    coin   = { icon = "Interface\\Icons\\INV_Misc_Coin_02", tint = { 1.0, 0.9, 0.4 }, size = 16, raise = 4 },
    chest  = { icon = "Interface\\Icons\\INV_Box_01", tint = { 0.9, 0.7, 0.3 }, size = 26, raise = 8 },
    bomb   = { icon = "Interface\\Icons\\INV_Misc_Bomb_02", tint = { 1.0, 0.5, 0.3 }, size = 22, raise = 10 },
    stone  = { icon = "Interface\\Icons\\Spell_Nature_EarthBind", tint = { 0.5, 0.9, 1.0 }, size = 20, raise = 10 },
    -- Supply caches appear on the field over time (the WaveManager places
    -- them), holding a random prize - worth walking off the beaten path for.
    cache  = { icon = "Interface\\Icons\\INV_Crate_01", tint = { 1.0, 0.9, 0.6 }, size = 28, raise = 8 },
    -- Hourglass of the Bronze: freezes every enemy (but Death) for 8 seconds.
    hourglass = { icon = "Interface\\Icons\\INV_Misc_PocketWatch_01", tint = { 0.6, 0.85, 1.0 }, size = 24, raise = 10 },
    -- Destinations you walk to (they do not drift toward you).
    merchant = { icon = "Interface\\Icons\\INV_Misc_Bag_10", tint = { 1.0, 0.9, 0.5 }, size = 30, noMagnet = true, raise = 10 },
    coffin   = { icon = "Interface\\Icons\\INV_Misc_Urn_01", tint = { 0.85, 0.85, 0.95 }, size = 34, noMagnet = true, raise = 10 },
    runeblade = { icon = "Interface\\Icons\\INV_Sword_121", tint = { 0.60, 1.00, 0.55 }, size = 36, noMagnet = true, raise = 10 },
    warglaives = { icon = "Interface\\Icons\\INV_Weapon_Glave_01", tint = { 0.70, 0.35, 1.00 }, size = 36, noMagnet = true, raise = 10 },
}

function Pickup:Initialize()
    self.pool = WS.Pool:New(function()
        return { frame = WS.UI:CreateEntityFrame(20) }
    end, function(pickup) pickup.frame:Hide() end, 12)
end

-- Free a slot when the pickup field is full by dropping the farthest COIN - the
-- only expendable pickup. Everything else (potions, bombs, lodestones,
-- hourglasses, caches, chests) is preserved. Returns true if a slot is free.
function Pickup:MakeRoom()
    if self.pool.count < WS.Constants.MAX_PICKUPS then return true end
    local player = WS.Game.player
    local best, bestD
    for i = 1, self.pool.count do
        local p = self.pool.active[i]
        if p.kind == "coin" then
            local d = player and WS.DistanceSquared(p.x, p.y, player.x, player.y) or 0
            if not bestD or d > bestD then bestD, best = d, p end
        end
    end
    if best then self.pool:Release(best); return true end
    return false
end

-- Any spawn that would fail on a full field first bumps the farthest coin to make
-- room; it's only dropped if there isn't even a coin to sacrifice.
function Pickup:Spawn(kind, x, y, value)
    if self.pool.count >= WS.Constants.MAX_PICKUPS and not self:MakeRoom() then return end
    local style = KINDS[kind]
    local pickup = self.pool:Acquire()
    pickup.kind, pickup.value = kind, value or 0
    pickup.x = WS.Clamp(x, 20, WS.Constants.WORLD_WIDTH - 20)
    pickup.y = WS.Clamp(y, 20, WS.Constants.WORLD_HEIGHT - 20)
    pickup.radius = style.noMagnet and 18 or 10
    pickup.size = style.size
    pickup.noMagnet = style.noMagnet
    pickup.frame.icon:SetTexture(style.icon)
    pickup.frame.icon:SetVertexColor(style.tint[1], style.tint[2], style.tint[3])
    pickup.frame:SetFrameLevel(WS.UI.world:GetFrameLevel() + (style.raise or 4))
    pickup.frame:Show()
    WS.UI:Place(pickup.frame, pickup.x, pickup.y, style.size)
    return pickup -- nil when the pool was full, so callers can skip the announce
end

-- Death-roll: called for every kill. Elites and bosses always leave a chest;
-- everything else rolls small chances scaled by the survivor's luck.
function Pickup:OnKill(enemy)
    local player = WS.Game.player
    local map = WS.Game.map
    if enemy.boss or enemy.elite then
        self:Spawn("chest", enemy.x, enemy.y, WS.floor((18 + WS.random(0, 26)) * WS.Game.run.goldMult))
        return
    end
    local luck = player.luck
    local roll = WS.random()
    local config = WS.Config
    if roll < config.dropChanceGold * luck then
        self:Spawn("coin", enemy.x, enemy.y, WS.floor((3 + WS.random(0, 4)) * WS.Game.run.goldMult))
    elseif roll < (config.dropChanceGold + config.dropChancePotion) * luck then
        self:Spawn("potion", enemy.x, enemy.y)
    elseif roll < (config.dropChanceGold + config.dropChancePotion + config.dropChanceBomb) * luck then
        self:Spawn("bomb", enemy.x, enemy.y)
    elseif roll < (config.dropChanceGold + config.dropChancePotion + config.dropChanceBomb + config.dropChanceStone) * luck then
        self:Spawn("stone", enemy.x, enemy.y)
    elseif roll < (config.dropChanceGold + config.dropChancePotion + config.dropChanceBomb + config.dropChanceStone + config.dropChanceHourglass) * luck then
        self:Spawn("hourglass", enemy.x, enemy.y)
    end
end

------------------------------------------------------------------------------
-- Effects on pickup
------------------------------------------------------------------------------

-- A goblin sapper charge: everything short of a boss dies screaming.
-- (Also triggered by lucky supply caches.)
function Pickup.ApplyBomb(player)
    WS.Audio:Play("bomb")
    WS.UI:Shake(10, 0.5)
    WS.FloatingText:Notice(player.x, player.y, "KABOOM!", 1, 0.5, 0.2)
    local enemies = WS.Enemy.pool
    local i = 1
    while i <= enemies.count do
        local enemy = enemies.active[i]
        if not enemy.boss then
            WS.Enemy:Damage(enemy, enemy.health + 1, false, "bomb")
            if enemy.poolIndex then i = i + 1 end
        else
            i = i + 1
        end
    end
end

local function Apply(pickup, player)
    local kind = pickup.kind
    if kind == "potion" then
        WS.Player:Heal(player, WS.floor(player.maxHealth * (WS.Config.potionHealPct or 0.30)), "potion")
        WS.Audio:Play("potion")

    elseif kind == "coin" then
        WS.Game:AddGold(WS.floor(pickup.value * player.goldMultiplier), pickup.x, pickup.y)

    elseif kind == "chest" then
        -- Jackpot ceremony: usually one bundle, sometimes three, rarely five.
        local roll = WS.random()
        local rolls = (roll < 0.05 and 5) or (roll < 0.30 and 3) or 1
        local total = WS.floor(pickup.value * player.goldMultiplier * (rolls == 1 and 1 or rolls * 0.8))
        WS.Game:AddGold(total, pickup.x, pickup.y)
        WS.UI:ChestCeremony(rolls, total)

    elseif kind == "bomb" then
        Pickup.ApplyBomb(player)

    elseif kind == "stone" then
        -- Lodestone: every gem on the field flies to the survivor.
        WS.XP:VacuumAll()
        WS.FloatingText:Notice(player.x, player.y, "Lodestone!", 0.5, 0.9, 1)
        WS.Audio:Play("gem")

    elseif kind == "cache" then
        -- A supply crate always holds one of four useful drops: a bomb, a
        -- lodestone, an hourglass, or two potions - never gold or a dud.
        WS.Audio:Play("chest")
        local roll = WS.random(1, 4)
        if roll == 1 then
            Pickup.ApplyBomb(player)
        elseif roll == 2 then
            WS.XP:VacuumAll()
            WS.FloatingText:Notice(player.x, player.y, "Lodestone!", 0.5, 0.9, 1)
            WS.Audio:Play("gem")
        elseif roll == 3 then
            WS.Enemy:FreezeAll(WS.Config.hourglassFreeze or 8)
            WS.FloatingText:Notice(player.x, player.y, "Time Stands Still!", 0.6, 0.85, 1)
            WS.Projectile:SpawnFlash(player.x, player.y, 200, WS.Constants.COLORS.frost)
            WS.Audio:Play("freeze")
        else
            WS.Player:Heal(player, WS.floor(player.maxHealth * (WS.Config.potionHealPct or 0.30) * 2), "potion")
            WS.FloatingText:Notice(player.x, player.y, "Two potions!", 0.5, 1, 0.4)
            WS.Audio:Play("potion")
        end

    elseif kind == "hourglass" then
        -- The Bronze Dragonflight stops time for the horde.
        WS.Enemy:FreezeAll(WS.Config.hourglassFreeze or 8)
        WS.FloatingText:Notice(player.x, player.y, "Time Stands Still!", 0.6, 0.85, 1)
        WS.Projectile:SpawnFlash(player.x, player.y, 200, WS.Constants.COLORS.frost)
        WS.Audio:Play("freeze")

    elseif kind == "merchant" then
        -- A wandering merchant hands you a free boon.
        WS.FloatingText:Notice(player.x, player.y, "A gift, friend!", 1, 0.9, 0.5)
        WS.Audio:Play("chest")
        WS.Game:GrantFreeBoon()

    elseif kind == "coffin" then
        -- The weathered coffin of an adequate knight.
        WoWSurvivorsDB.statistics.coffinsOpened = (WoWSurvivorsDB.statistics.coffinsOpened or 0) + 1
        WS.Audio:Play("evolve")
        WS.UI:Shake(5, 0.4)
        WS.Projectile:SpawnFlash(player.x, player.y, 120, WS.Constants.COLORS.holy)
        if not WoWSurvivorsDB.unlocks.characters.paladin then
            WoWSurvivorsDB.unlocks.characters.paladin = true
            WS.UI:Toast(WS.Characters.paladin.classIcon, "A Survivor Rises!", WS.Characters.paladin.name)
            WS.Game:Announce(WS.Characters.paladin.name .. " is freed!", "A new survivor joins the roster.", 4.0)
        else
            WS.Game:AddGold(WS.floor(80 * WS.Game.run.goldMult), player.x, player.y)
        end
        WS.Achievements:Check()

    elseif kind == "runeblade" then
        -- The blade that answered the rot. Claiming it frees the Death Knight.
        WoWSurvivorsDB.statistics.runebladesClaimed =
            (WoWSurvivorsDB.statistics.runebladesClaimed or 0) + 1
        WS.Audio:Play("evolve")
        WS.UI:Shake(7, 0.5)
        WS.Projectile:SpawnFlash(player.x, player.y, 140, { 0.45, 1.0, 0.40 })
        if not WoWSurvivorsDB.unlocks.characters.death_knight then
            WoWSurvivorsDB.unlocks.characters.death_knight = true
            WS.UI:Toast(WS.Characters.death_knight.classIcon, "The Blade Claims You",
                WS.Characters.death_knight.name)
            WS.Game:Announce(WS.Characters.death_knight.name .. " rises.",
                "What the Light could not hold, the blade keeps.", 4.0)
        else
            WS.Game:AddGold(WS.floor(120 * WS.Game.run.goldMult), player.x, player.y)
        end
        WS.Achievements:Check()

    elseif kind == "warglaives" then
        -- The Warglaives of Azzinoth's lesser cousins. Claiming them frees the DH.
        WoWSurvivorsDB.statistics.warglaivesClaimed =
            (WoWSurvivorsDB.statistics.warglaivesClaimed or 0) + 1
        WS.Audio:Play("evolve")
        WS.UI:Shake(7, 0.5)
        WS.Projectile:SpawnFlash(player.x, player.y, 150, { 0.60, 1.0, 0.25 })
        if not WoWSurvivorsDB.unlocks.characters.demon_hunter then
            WoWSurvivorsDB.unlocks.characters.demon_hunter = true
            WS.UI:Toast(WS.Characters.demon_hunter.classIcon, "You Are Prepared",
                WS.Characters.demon_hunter.name)
            WS.Game:Announce(WS.Characters.demon_hunter.name .. " takes up the glaives.",
                "The fel was always going to win.", 4.0)
        else
            WS.Game:AddGold(WS.floor(120 * WS.Game.run.goldMult), player.x, player.y)
        end
        WS.Achievements:Check()
    end
end

------------------------------------------------------------------------------

function Pickup:Update(dt)
    local player = WS.Game.player
    local i = 1
    while i <= self.pool.count do
        local pickup = self.pool.active[i]
        local dx, dy, distance = WS.Normalize(player.x - pickup.x, player.y - pickup.y)
        -- Destination pickups (coffin, merchant) never drift; you walk to them.
        if not pickup.noMagnet and distance < player.pickupRadius then
            local pull = WS.max(160, 420 * (1 - distance / player.pickupRadius))
            pickup.x = pickup.x + dx * pull * dt
            pickup.y = pickup.y + dy * pull * dt
        end
        if distance < player.radius + pickup.radius + 6 then
            Apply(pickup, player)
            self.pool:Release(pickup)
            if not WS.Game.running then return end
        else
            WS.UI:Place(pickup.frame, pickup.x, pickup.y, pickup.size)
            i = i + 1
        end
    end
end

function Pickup:Clear()
    if self.pool then self.pool:ReleaseAll() end
end

File: src/game/pickup.js
/* Ground pickups: coins, potions, chests, sapper charges, lodestones, bronze
 * hourglasses, supply caches, and the two story objects (the coffin and the
 * egg merchant). Everything drifts toward the survivor once inside pickup
 * range, except destination pickups you have to walk to. */
'use strict';
(function (WS) {

  const TYPES = {
    coin: { art: 'coin', tint: [1.0, 0.82, 0.20], size: 15 },
    potion: { art: 'potion', tint: [1.0, 0.40, 0.40], size: 17 },
    chest: { art: 'chest', tint: [0.9, 0.70, 0.30], size: 22 },
    bomb: { art: 'bomb', tint: [1.0, 0.50, 0.30], size: 19 },
    stone: { art: 'stone', tint: [0.5, 0.90, 1.0], size: 18 },
    cache: { art: 'cache', tint: [1.0, 0.90, 0.60], size: 24 },
    hourglass: { art: 'hourglass', tint: [0.6, 0.85, 1.0], size: 20 },
    coffin: { art: 'coffin', tint: [0.85, 0.85, 0.95], size: 30, noMagnet: true },
    graveblade: { art: 'graveblade', tint: [0.85, 0.20, 0.25], size: 30, noMagnet: true },
    twinglaive: { art: 'twinglaive', tint: [0.55, 1.0, 0.25], size: 30, noMagnet: true },
    merchant: { art: 'merchant', tint: [0.95, 0.85, 0.45], size: 30, noMagnet: true },
  };

  const Pickup = { pool: null, TYPES };

  Pickup.init = function () {
    this.pool = new WS.Pool(() => ({}), null, WS.CONST.MAX_PICKUPS);
  };

  Pickup.clear = function () { if (this.pool) this.pool.releaseAll(); };

  /** Coins are the only expendable pickup; everything else is preserved when
   *  the field fills up. Returns the pickup, or null if there was no room. */
  Pickup.spawn = function (kind, x, y, value) {
    if (this.pool.count >= WS.CONST.MAX_PICKUPS) {
      let coin = -1;
      for (let i = 0; i < this.pool.count; i++) {
        if (this.pool.active[i].kind === 'coin') { coin = i; break; }
      }
      if (coin < 0) return null;
      this.pool.releaseAt(coin);
    }
    const p = this.pool.acquire();
    if (!p) return null;
    p.kind = kind;
    p.type = TYPES[kind];
    p.x = x; p.y = y;
    p.value = value || 0;
    p.radius = p.type.size * 0.5;
    p.bob = WS.random() * WS.TAU;
    p.life = 0;
    p.snooze = 0;
    return p;
  };

  /** Death-roll, called for every kill. Elites and bosses always leave a chest. */
  Pickup.onKill = function (enemy) {
    const player = WS.Game.player;
    const cfg = WS.Config;
    const luck = player.luck;

    if (enemy.elite) {
      this.spawn('chest', enemy.x, enemy.y, WS.floor((18 + WS.randInt(0, 26)) * WS.Game.run.goldMult));
      return;
    }
    if (enemy.boss) return;   // bosses pay out directly in Enemy.kill

    if (WS.random() < cfg.dropChanceGold * luck) {
      this.spawn('coin', enemy.x, enemy.y, WS.floor((3 + WS.randInt(0, 4)) * WS.Game.run.goldMult));
    } else if (WS.random() < cfg.dropChancePotion * luck) {
      this.spawn('potion', enemy.x, enemy.y);
    } else if (WS.random() < cfg.dropChanceBomb * luck) {
      this.spawn('bomb', enemy.x, enemy.y);
    } else if (WS.random() < cfg.dropChanceStone * luck) {
      this.spawn('stone', enemy.x, enemy.y);
    } else if (WS.random() < cfg.dropChanceHourglass * luck) {
      this.spawn('hourglass', enemy.x, enemy.y);
    }
  };

  /** Goblin sapper charge: clears the screen. Never one-shots a boss. */
  Pickup.detonate = function (x, y) {
    WS.Audio.play('explode');
    WS.FX.shake(9, 0.5);
    WS.FX.flash(x, y, 520, [1.0, 0.65, 0.25], 0.55);
    WS.FX.screen('rgba(255,180,80,.22)', 0.35);
    const enemies = WS.Enemy.pool.active;
    for (let i = enemies.length - 1; i >= 0; i--) {
      const e = enemies[i];
      if (e.boss) WS.Enemy.hit(e, e.maxHealth * 0.08, 'bomb');
      else WS.Enemy.damage(e, e.health + 1, false, 'bomb');
    }
  };

  function collect(p, pickup) {
    const kind = pickup.kind;
    const run = WS.Game.run;

    if (kind === 'potion') {
      WS.Player.heal(p, WS.floor(p.maxHealth * WS.Config.potionHealPct), 'potion');
      WS.Audio.play('potion');

    } else if (kind === 'coin') {
      WS.Game.addGold(WS.floor(pickup.value * p.goldMultiplier), pickup.x, pickup.y);
      WS.Audio.play('coin');

    } else if (kind === 'chest') {
      // Chests roll for a jackpot: one, three, or five payouts.
      const roll = WS.random();
      const rolls = roll < 0.03 ? 5 : roll < 0.15 ? 3 : 1;
      const total = WS.floor(pickup.value * p.goldMultiplier * (rolls === 1 ? 1 : rolls * 0.8));
      WS.Game.addGold(total, pickup.x, pickup.y);
      WS.Audio.play('chest');
      if (rolls > 1) {
        WS.Game.toast(rolls === 5 ? 'JACKPOT!' : 'Treasure!', `The chest pays out ${rolls}x.`);
        WS.FX.flash(pickup.x, pickup.y, 90, WS.CONST.COLORS.gold, 0.5);
      }

    } else if (kind === 'bomb') {
      Pickup.detonate(pickup.x, pickup.y);

    } else if (kind === 'stone') {
      WS.XP.vacuumAll();
      WS.FX.notice(p.x, p.y, 'Lodestone!', '#8fe0ff');
      WS.Audio.play('gem');

    } else if (kind === 'hourglass') {
      WS.Enemy.freezeAll(WS.Config.hourglassFreeze);
      WS.FX.notice(p.x, p.y, 'Time stops!', '#bfe6ff');

    } else if (kind === 'cache') {
      // A supply crate always holds something useful - never gold, never a dud.
      WS.Audio.play('chest');
      const roll = WS.randInt(1, 4);
      if (roll === 1) Pickup.detonate(pickup.x, pickup.y);
      else if (roll === 2) { WS.XP.vacuumAll(); WS.FX.notice(p.x, p.y, 'Lodestone!', '#8fe0ff'); }
      else if (roll === 3) { WS.Enemy.freezeAll(WS.Config.hourglassFreeze); WS.FX.notice(p.x, p.y, 'Time stops!', '#bfe6ff'); }
      else {
        WS.Player.heal(p, WS.floor(p.maxHealth * WS.Config.potionHealPct * 2), 'potion');
        WS.FX.notice(p.x, p.y, 'Two potions!', '#7fe89a');
        WS.Audio.play('potion');
      }

    } else if (kind === 'coffin') {
      WS.Save.stats.coffinsOpened = (WS.Save.stats.coffinsOpened || 0) + 1;
      WS.Save.save();
      WS.Game.announce('Bartholomew the Adequate rises!',
        'He files a complaint, then hands you the shield.', 3.5);
      WS.Game.addGold(WS.floor(80 * run.goldMult), p.x, p.y);
      WS.Achievements.check();
      WS.Audio.play('evolve');

    } else if (kind === 'graveblade') {
      WS.Save.stats.gravebladesClaimed = (WS.Save.stats.gravebladesClaimed || 0) + 1;
      WS.Save.save();
      WS.Game.announce('The graveblade answers.', 'Something colder takes the hilt.', 3.5,
        { kind: 'glory' });
      WS.Game.addGold(WS.floor(120 * run.goldMult), p.x, p.y);
      WS.Player.addWeapon(p, 'reaving_arc');
      WS.Achievements.check();
      WS.Audio.play('evolve');

    } else if (kind === 'twinglaive') {
      WS.Save.stats.glaivesClaimed = (WS.Save.stats.glaivesClaimed || 0) + 1;
      WS.Save.save();
      WS.Game.announce('The twin glaives find you.', 'You were never going to refuse.', 3.5,
        { kind: 'glory' });
      WS.Game.addGold(WS.floor(120 * run.goldMult), p.x, p.y);
      WS.Player.addWeapon(p, 'verdant_lance');
      WS.Achievements.check();
      WS.Audio.play('evolve');

    } else if (kind === 'merchant') {
      // The Egg Merchant sells run-only eggs; "Buy All" spends every coin.
      const cost = WS.Config.eggVendorCost;
      const affordable = WS.floor(run.gold / cost);
      if (affordable <= 0) {
        WS.Game.toast('Egg Merchant', `You cannot afford an egg (${cost}g each).`);
        return false;   // leave the merchant standing
      }
      run.gold -= affordable * cost;
      WS.Player.grantRunEggs(p, affordable);
      WS.Game.toast('Egg Merchant', `Bought ${affordable} curious egg${affordable === 1 ? '' : 's'}.`);
      WS.Audio.play('chest');
    }
    return true;
  }

  Pickup.update = function (dt) {
    const player = WS.Game.player;
    let i = 0;
    while (i < this.pool.count) {
      const pickup = this.pool.active[i];
      pickup.bob += dt * 3;
      pickup.life += dt;
      const [dx, dy, distance] = WS.normalize(player.x - pickup.x, player.y - pickup.y);

      if (!pickup.type.noMagnet && distance < player.pickupRadius) {
        const pull = WS.max(150, 380 * (1 - distance / player.pickupRadius));
        pickup.x += dx * pull * dt;
        pickup.y += dy * pull * dt;
      }

      if (pickup.snooze > 0) pickup.snooze -= dt;
      if (pickup.snooze <= 0 && distance < player.radius + pickup.radius + 4) {
        const consumed = collect(player, pickup);
        if (consumed) {
          this.pool.releaseAt(i);
          if (!WS.Game.running) return;
          continue;
        }
        // Not consumed (the merchant, with no coin to spend): wait a few
        // seconds before offering again instead of toasting every frame.
        pickup.snooze = 4;
      }
      i++;
    }
  };

  WS.Pickup = Pickup;

})(window.WS);
